src/crypto/variant4_random_math.h

```diff
@@ -0,0 +1,405 @@
+	// I've checked all block heights < 10,000,000 and here is the distribution of program sizes:
+	//
+	// 60      28495
+	// 61      106077
+	// 62      2455855
+	// 63      5114930
+	// 64      1020868
+	// 65      1109026
+	// 66      151756
+	// 67      8429
+	// 68      4477
+	// 69      87
+
+	// Unroll 70 instructions here
+	V4_EXEC_10(0);		// instructions 0-9
+	V4_EXEC_10(10);		// instructions 10-19
+	V4_EXEC_10(20);		// instructions 20-29
+	V4_EXEC_10(30);		// instructions 30-39
+	V4_EXEC_10(40);		// instructions 40-49
+	V4_EXEC_10(50);		// instructions 50-59
+	V4_EXEC_10(60);		// instructions 60-69
+
+#undef V4_EXEC_10
+#undef V4_EXEC
+}
+
+// If we don't have enough data available, generate more
+static FORCEINLINE void check_data(size_t* data_index, const size_t bytes_needed, char* data, const size_t data_size)
+{
+	if (*data_index + bytes_needed > data_size)
+	{
+		hash_extra_blake(data, data_size, data);
+		*data_index = 0;
+	}
+}
+
+// Generates as many random math operations as possible with given latency and ALU restrictions
+static inline int v4_random_math_init(struct V4_Instruction* code, const uint64_t height)
+{
+	// MUL is 3 cycles, 3-way addition and rotations are 2 cycles, SUB/XOR are 1 cycle
+	// These latencies match real-life instruction latencies for Intel CPUs starting from Sandy Bridge and up to Skylake/Coffee lake
+	//
+	// AMD Ryzen has the same latencies except 1-cycle ROR/ROL, so it'll be a bit faster than Intel Sandy Bridge and newer processors
+	// Surprisingly, Intel Nehalem also has 1-cycle ROR/ROL, so it'll also be faster than Intel Sandy Bridge and newer processors
+	// AMD Bulldozer has 4 cycles latency for MUL (slower than Intel) and 1 cycle for ROR/ROL (faster than Intel), so average performance will be the same
+	// Source: https://www.agner.org/optimize/instruction_tables.pdf
+	const int op_latency[V4_INSTRUCTION_COUNT] = { 3, 2, 1, 2, 2, 1 };
+
+	// Instruction latencies for theoretical ASIC implementation
+	const int asic_op_latency[V4_INSTRUCTION_COUNT] = { 3, 1, 1, 1, 1, 1 };
+
+	// Available ALUs for each instruction
+	const int op_ALUs[V4_INSTRUCTION_COUNT] = { ALU_COUNT_MUL, ALU_COUNT, ALU_COUNT, ALU_COUNT, ALU_COUNT, ALU_COUNT };
+
+	char data[32];
+	memset(data, 0, sizeof(data));
+	*((uint64_t*)data) = height;
+
+	size_t data_index = sizeof(data);
+
+	int code_size;
+	do {
+		int latency[8];
+		int asic_latency[8];
+
+		// Tracks previous instruction and value of the source operand for registers R0-R3 throughout code execution
+		// byte 0: current value of the destination register
+		// byte 1: instruction opcode
+		// byte 2: current value of the source register
+		//
+		// Registers R4-R7 are constant and are threatened as having the same value because when we do
+		// the same operation twice with two constant source registers, it can be optimized into a single operation
+		int inst_data[8] = { 0, 1, 2, 3, -1, -1, -1, -1 };
```

Comment:
- No, the outer `do` loop is for generating the whole program. It runs more than once only if the first generated program is too short.
- Initial values must be different for R0-R3, this is the only requirement. I should probably fix them to enable any instruction as first
- Will change to uint32_t
- Yes, this is by design, see comment above

Reply:
> Initializing to `-1` has implementation defined bit values - the type should likely be `unsigned` to prevent (unlikely) portability issues.

On the second thought - do you even know modern CPU architecture which doesn't use two's complement representation for negative numbers?

Reply:
I'm reasonably sure this code would not work on non 2's complement anyway, given it subs them rols, etc.

Reply:
There's a proposal to mandate that integers are twos-complement in C++
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0907r4.html

I don't think there's a corresponding proposal for C yet. There are still ones-complement machines in existence, but I don't think we'll ever see anyone using them for Monero nodes. (https://stackoverflow.com/questions/12276957/are-there-any-non-twos-complement-implementations-of-c)

Reply:
> * No, the outer `do` loop is for generating the whole program. It runs more than once only if the first generated program is too short.

This didn't answer my original question. I know what the `do while` loop was doing, I wanted to know why tracking prior instructions was skipped in some situations.

> On the second thought - do you even know modern CPU architecture which doesn't use two's complement representation for negative numbers?

I'm not sure what it matters, making the type `unsigned` is portable now and doesn't affect anything else in the code.

Reply:
> This didn't answer my original question. I know what the `do while` loop was doing, I wanted to know why tracking prior instructions was skipped in some situations.

The answer to your original question: no, `inst_data` shouldn't be declared before the `do` loop. I don't want previous generated program that was scrapped for some reason to influence the next attempt.

Reply:
> This might need to be `uint32` to be completely portable now that I think about it. Although, that meant `int` had two portability issues actually.

What portability issues?
- Two's complement: not an issue in reality
- Signed overflow: look at the code and see that it never happens.

Don't be picky.

Reply:
I've made it uint32_t since it doesn't change the code logic anyway, but can you elaborate? Maybe I don't know C/C++ standard well enough.

Reply:
The standard states that `int`/`unsigned` must be 16- bits+ (not sure if there is cap) and does not specify binary format for negative numbers.

Reply:
As noted earlier https://github.com/monero-project/monero/pull/5126#discussion_r255299070 the case for other-than-twos-complement isn't worth worrying about.